image/refs/interface.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package refs implements generic name-based reference access.
+package refs
+
+import (
+	"github.com/opencontainers/image-spec/image/structure"
+)
+
+// Engine represents a name-based reference storage engine.
+type Engine interface {
+
+	// Put adds a new reference to the store.
+	Put(name string, descriptor *structure.Descriptor) (err error)
+
+	// Get returns a reference from the store.
+	Get(name string) (descriptor *structure.Descriptor, err error)
```

Comment:
On Fri, Jul 01, 2016 at 09:04:29AM -0700, Simone Gotti wrote:
> > +	// Get returns a reference from the store.
> > +	Get(name string) (descriptor *structure.Descriptor, err error)
> 
> If the reference doesn't exists, descriptor will be nil or an error
> will be returned?

Ah, good point.  Currently this returns io.EOF as an error [1,2]
because that's what tar does [3].  But that's probably not an
intuitive error for a more generic “key not found” problem.  Is there
a stock Go error that's a better match, or should we create our own?
If we create our own, do we want separate error variables for the cas
and refs interfaces?

And however this goes down, we'll want to document the key-not-found
case here.

[1]: https://github.com/opencontainers/image-spec/pull/159/files#diff-b11cc21c1cf82254929b0e2e98cbd695R74
[2]: https://github.com/opencontainers/image-spec/pull/159/files#diff-ce780f6bc5726d182fefbd55676f47eaR70
[3]: https://golang.org/pkg/archive/tar/#Reader.Next